libtrellis/tools/ecppll.cpp

```diff
@@ -0,0 +1,176 @@
+  options.add_options()("input,i", po::value<float>()->required(), "Input frequency in MHz");
+  options.add_options()("output,o", po::value<float>()->required(), "Output frequency in MHz");
+  options.add_options()("file,f", po::value<string>(), "Output to file");
+
+  po::variables_map vm;
+  po::parsed_options parsed = po::command_line_parser(argc, argv).options(options).run();
+  po::store(parsed, vm);
+  po::notify(vm);
+
+  if(vm.count("help")){
+    cerr << "Project Trellis - Open Source Tools for ECP5 FPGAs" << endl;
+    cerr << "ecpunpack: ECP5 bitstream to text config converter" << endl;
+    cerr << endl;
+    cerr << "Copyright (C) 2018 David Shah <david@symbioticeda.com>" << endl;
+    cerr << endl;
+    cerr << options << endl;
+  }
+  float inputf = vm["input"].as<float>();
+  float outputf = vm["output"].as<float>();
+  if(inputf < INPUT_MIN || inputf > INPUT_MAX){
+    fprintf(stderr, "Input frequency %.3fMHz not in range (%.3fMHz, %.3fMHz)\n",
+	    inputf, INPUT_MIN, INPUT_MAX);
+    return 1;
+  }
+  if(outputf < OUTPUT_MIN || outputf > OUTPUT_MAX){
+    fprintf(stderr, "Output frequency %.3fMHz not in range (%.3fMHz, %.3fMHz)\n",
+	    outputf, OUTPUT_MIN, OUTPUT_MAX);
+    return 1;
+  }
+  pll_params params = calc_pll_params(inputf, outputf);
+
+  fprintf(stdout, "Pll parameters:\n");
+  fprintf(stdout, "Refclk divisor: %d\n", params.refclk_div);
+  fprintf(stdout, "Feedback divisor: %d\n", params.feedback_div);
+  fprintf(stdout, "Output divisor: %d\n", params.output_div);
+  fprintf(stdout, "VCO frequency: %f\n", params.fvco);
+  fprintf(stdout, "Output frequency: %f\n", params.fout);
+  if(vm.count("file")){
+    FILE *f = NULL;
+
+    if(vm["file"].as<string>() == "-")
+      f = stdout;
+    else
+      f = fopen(vm["file"].as<string>().c_str(), "w");
+
+    
+    write_pll_config(params, "pll", f);
+
+  }
+
+}
+
+pll_params calc_pll_params(float input, float output){
+  float error = std::numeric_limits<float>::max();
+  pll_params params = {0};
+  for(int input_div=1;input_div <= 128; input_div++){
+
+    float fpfd = input / (float)input_div;
+    if(fpfd < PFD_MIN || fpfd > PFD_MAX)
+      continue;
+    for(int feedback_div=1;feedback_div <= 80; feedback_div++){
+      for(int output_div=1;output_div <= 128; output_div++){
+	float fvco = fpfd * (float)feedback_div * (float) output_div;
+	
+	if(fvco < VCO_MIN || fvco > VCO_MAX)
+	  continue;
+
+	float fout = fvco / (float) output_div;
+	if(fabsf(fout - output) < error ||
+	   (fabsf(fout-output) == error && fabsf(fvco - 600) < fabsf(params.fvco - 600))){
+	  error = fabsf(fout-output);
+	  params.refclk_div = input_div;
+	  params.feedback_div = feedback_div;
+	  params.output_div = output_div;
+	  params.fout = fout;
+	  params.fvco = fvco;
+
+	}
+
+      }
+    }
+  }
+  return params;
+}
+
+void write_pll_config(pll_params params, const char* name,  FILE* file){
+  fprintf(file, "module %s(input clki, output clko);\n", name);
```

Comment:
I don't have a problem merging this like this, but in general  the Trellis convention has been to use C++ libraries (fstream/iostream) over C.